pkgs.rosPackages.humble.cartographer-ros: Build fail

`cartographer-ros` under the `humble` distribution fails to build when declared. See a truncated log snippet below:

```
building '/nix/store/kvfyf4qdw4nvfb7fb35fnrkdpai0n2gw-ros-humble-cartographer-ros-2.0.9002-r1.drv'...
error: builder for '/nix/store/kvfyf4qdw4nvfb7fb35fnrkdpai0n2gw-ros-humble-cartographer-ros-2.0.9002-r1.drv' failed with exit code 2;
       last 25 log lines:
       > /build/cartographer_ros-release-release-humble-cartographer_ros-2.0.9002-1/include/cartographer_ros/node.h:171:24: error: expected ';' at end of member declaration
       >   171 |       int trajectory_id) EXCLUSIVE_LOCKS_REQUIRED(mutex_);
       >       |                        ^
       >       |                         ;
       > /build/cartographer_ros-release-release-humble-cartographer_ros-2.0.9002-1/include/cartographer_ros/node.h:171:51: error: 'mutex_' has not been declared
       >   171 |       int trajectory_id) EXCLUSIVE_LOCKS_REQUIRED(mutex_);
       >       |                                                   ^~~~~~
       > /build/cartographer_ros-release-release-humble-cartographer_ros-2.0.9002-1/include/cartographer_ros/node.h:171:26: error: ISO C++ forbids declaration of 'EXCLUSIVE_LOCKS_REQUIRED' with no type []
       >   171 |       int trajectory_id) EXCLUSIVE_LOCKS_REQUIRED(mutex_);
       >       |                          ^~~~~~~~~~~~~~~~~~~~~~~~
       > /build/cartographer_ros-release-release-humble-cartographer_ros-2.0.9002-1/include/cartographer_ros/node.h:186:37: error: expected ';' at end of member declaration
       >   186 |   std::shared_ptr<MapBuilderBridge> map_builder_bridge_ GUARDED_BY(mutex_);
       >       |                                     ^~~~~~~~~~~~~~~~~~~
       >       |                                                        ;
       > /build/cartographer_ros-release-release-humble-cartographer_ros-2.0.9002-1/include/cartographer_ros/node.h:186:68: error: 'mutex_' is not a type
       >   186 |   std::shared_ptr<MapBuilderBridge> map_builder_bridge_ GUARDED_BY(mutex_);
       >       |                                                                    ^~~~~~
       > /build/cartographer_ros-release-release-humble-cartographer_ros-2.0.9002-1/include/cartographer_ros/node.h:186:57: error: ISO C++ forbids declaration of 'GUARDED_BY' with no type []
       >   186 |   std::shared_ptr<MapBuilderBridge> map_builder_bridge_ GUARDED_BY(mutex_);
       >       |                                                         ^~~~~~~~~~
       > make[2]: *** [CMakeFiles/cartographer_ros.dir/build.make:93: CMakeFiles/cartographer_ros.dir/src/map_builder_bridge.cpp.o] Error 1
       > make[2]: *** [CMakeFiles/cartographer_ros.dir/build.make:163: CMakeFiles/cartographer_ros.dir/src/offline_node.cpp.o] Error 1
       > make[2]: *** [CMakeFiles/cartographer_ros.dir/build.make:135: CMakeFiles/cartographer_ros.dir/src/node.cpp.o] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:174: CMakeFiles/cartographer_ros.dir/all] Error 2
       > make: *** [Makefile:146: all] Error 2
       For full logs, run 'nix log /nix/store/kvfyf4qdw4nvfb7fb35fnrkdpai0n2gw-ros-humble-cartographer-ros-2.0.9002-r1.drv'.
error: 1 dependencies of derivation '/nix/store/4i7igbb9d59b54ybnbcgmnfh660qqhy5-ros-env.drv' failed to build
error: 1 dependencies of derivation '/nix/store/cvc3yqb06qm28m6k2pagqvf145xx9lyn-ROS-2-Workspace-env.drv' failed to build
```


Using `develop` branch due to recently merged fixes for `cartographer`.

--- a/distros/ros2-overlay.nix
+++ b/distros/ros2-overlay.nix
@@ -69,6 +69,26 @@ rosSelf: rosSuper: with rosSelf.lib; {
     '';
   });
 
+  cartographer-ros = rosSuper.cartographer-ros.overrideAttrs ({
+    patches ? [],
+    postPatch ? "", ...
+  }: {
+    patches = patches ++ [
+      # Fix compilation with glog >= 0.7.0 (https://github.com/ros2/cartographer_ros/pull/76)
+      (self.fetchpatch {
+        url = "https://github.com/ros2/cartographer_ros/commit/58cd253615606efbf0bf69d16a932d35aadef1f7.patch";
+        hash = "sha256-rmolyUYIWPT37kYITDW4cRO0XJNdIYs6AoUWgWVb8PU=";
+        stripLen = 1;
+      })
+    ];
+
+    # Add ABSL_ prefix to thread annotation macros. See details above.
+    postPatch = ''
+      sed -i -Ee 's/\<(LOCKS_EXCLUDED|EXCLUSIVE_LOCKS_REQUIRED|GUARDED_BY)\>/ABSL_\1/g' \
+          $(find -name \*.h -o -name \*.cpp )
+    '';
+  });
+
   cyclonedds = rosSuper.cyclonedds.overrideAttrs ({
     cmakeFlags ? [], ...
   }: {

